:sparkles: Complete volume snapshot support and better snapshot management

mhkarimi1383 opened the pull request:
Resolves #19

mhkarimi1383 commented on rawfile/internal_svc.py:
```diff
@@ -42,30 +48,42 @@ def ExpandRawFile(
                 status=internal_pb2.ExpandRawFileStatus.OK,
             )
+
+    def GetRawFile(self, request: internal_pb2.GetRawFileRequest, context):
```
It will use a Temporary snapshot later on...

laurivosandi commented on rawfile/bd2fs.py:
```diff
@@ -285,20 +283,37 @@ def ControllerExpandVolume(self, request, context):
+            if volume_meta.get("freezefs", False) and mountpoints:
+                for mp in mountpoints:
+                    run(f"fsfreeze --freeze {mp}", check=True)
+            cmd = f"cp --sparse=always --reflink={'always' if volume_meta.get('copy_on_write', False) else 'never'}"
```
fsfreeze without reflinking would result in data corruption... convention over configuration...

mhkarimi1383 replied:
fs freeze only blocks write operations,

But if that's the case can you send some documents/test scenerios for that?
We can make this default and also warn users about that

But what about backing storage does not supper COW?

laurivosandi replied:
1. Reflinking should work without freezefs (XFS, Btrfs)
2. Freezefs could be used as fallback for other filesystems (ext4)
3. Anything else should explicitly fail. Why warn user if it will never work?

mhkarimi1383 replied:
Yeah,
We have decided to make these completely configurable

And make default options to be safe in every situation

But we will have a complete documentation on this and how options are working together

tiagolobocastro replied:
without fsfreeze or reflink we should reject snapshot if volume is in use

mhkarimi1383 replied:
Yeah 👍

mhkarimi1383 replied:
done. it is getting rejected right now

tiagolobocastro commented on rawfile/rawfile_servicer.py:
```diff
@@ -208,14 +223,19 @@ def CreateVolume(self, request, context):
         format_options = params.get("formatoptions", "").strip()
+        snap_inuse = str_to_bool(params.get("snapinuse", "no"))
+        freezefs = str_to_bool(params.get("freezefs", "no"))
+        copy_on_write = str_to_bool(params.get("copyonwrite", "no"))
```
shouldn't this come from filesystem where data is living on?

mhkarimi1383 replied:
You mean we should determine them from the data dir filesystem?

tiagolobocastro replied:
Yes

mhkarimi1383 replied:
I will check if we can determine that from underlining filesystem of data dir, So we can set copy on write automatically

mhkarimi1383 replied:
There is not way to check that without checking out FS and it's mount info

in xfs for example

```console
xfs_info /
meta-data=/dev/disk/by-uuid/72fcf69a-04b3-481a-9a94-2164a4c2d2b5 isize=512    agcount=4, agsize=62381653 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=1
         =                       reflink=1    bigtime=1 inobtcount=1 nrext64=1
         =                       exchange=0   metadir=0
data     =                       bsize=4096   blocks=249526609, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1, parent=0
log      =internal log           bsize=4096   blocks=121839, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
         =                       rgcount=0    rgsize=0 extents
         =                       zoned=0      start=0 reserved=0
```

`reflink=1` means I have COW, and in others we should check them by their FS too, or we should test reflinking before creating snapshots

tiagolobocastro replied:
So we could do this then?

mhkarimi1383 replied:
We can, but it is so dirty also users should be able to manage this with full control

tiagolobocastro replied:
This may become very difficult to manage though as this is a global value
I have seen clusters start very homogeneously, same nodes, memory, storage etc and as things evolve then different nodes get added, different kernel version, different OS, etc so it's very possible that not all nodes have the same COW filesystem?

mhkarimi1383 replied:
This is like what we have for LVM and others that we are settings a storage pool with some features enabled or disabled

If we add storage pool support we can get these params per pool and per node

tiagolobocastro replied:
We could make it configurable but I think by default we should figure out the correct value for cow based on the data filesystem; don't need to do it all at once though

mhkarimi1383 replied:
I will handle this :)

mhkarimi1383 replied:
Done.

tiagolobocastro commented on rawfile/rawfile_servicer.py:
```diff
@@ -208,14 +223,19 @@ def CreateVolume(self, request, context):
-        thin_provision = params.get("thinprovision", "no")
         format_options = params.get("formatoptions", "").strip()
+        snap_inuse = str_to_bool(params.get("snapinuse", "no"))
+        freezefs = str_to_bool(params.get("freezefs", "no"))
```
this should come from the snapshot class rather than volume?

mhkarimi1383 replied:
I have used storage class parameters because we will need them outside of the snapshots

tiagolobocastro replied:
Ah ok, where else would it be used then?

mhkarimi1383 replied:
for temporary snapshots for between node migration/clones
that's why I have set them in volume metadata

mhkarimi1383 replied:
I want to also add storage pools later on, So we will have these params there

tiagolobocastro replied:
Should freezefs be turned on by default?

mhkarimi1383 replied:
Default behavior should not block running applications

tiagolobocastro replied:
Shouldn't default behavior be safety first, a consistent snapshot? Other user may have nasty surprise later when trying to use it
So I think `dangerous` should be an opt-in

mhkarimi1383 replied:
We are already blocking attached snapshots

Both are OK if we document that

tiagolobocastro replied:
doh, forgot that we were already blocking `createSnapshot`, my bad!
Then I think we're good with that
We should also start documenting these new changes

mhkarimi1383 replied:
Yeah

any updates on product document itself?

tiagolobocastro commented on rawfile/rawfile_servicer.py:
```diff
@@ -208,14 +223,19 @@ def CreateVolume(self, request, context):
-        params = normalize_parameters(request.parameters)
-        thin_provision = params.get("thinprovision", "no")
         format_options = params.get("formatoptions", "").strip()
+        snap_inuse = str_to_bool(params.get("snapinuse", "no"))
```
what does this mean?

mhkarimi1383 replied:
This option controls if we are able to take snapshot from an in use volume or not

mhkarimi1383 replied:
We can also set this automatically based on COW and fsfreeze

tiagolobocastro replied:
yes :)

mhkarimi1383 replied:
done.

mhkarimi1383 commented:
I have also fixed size problem, it is not `0` anymore, and I have added `ListSnapshots` too and if snapshot size changes it will get reported

mhkarimi1383 commented:
@tiagolobocastro 
With this we can do snapshots in background I think :)

So if snapshot creation takes longer we won't get timeout errors

mhkarimi1383 commented:
@mergifyio rebase

tiagolobocastro commented on rawfile/bd2fs.py:
```diff
@@ -285,24 +285,49 @@ def ControllerExpandVolume(self, request, context):
-
-            if fs.mountpoint is None:
-                detach_loops(file)
+                return
```
pardon my ignorance, does this return, return the error?

mhkarimi1383 replied:
Yeah it is after context.abort

tiagolobocastro commented on rawfile/bd2fs.py:
```diff
@@ -285,24 +285,49 @@ def ControllerExpandVolume(self, request, context):
+                os.makedirs(snapshots_path, exist_ok=True)
+                if should_freeze:
+                    for mp in mountpoints:
+                        run(f"fsfreeze --freeze {mp}", check=True)
```
what happens if the pod is deleted whilst freezed, who does the unfreeze?
or I guess we always wait for create_snapshot to conclude on retry?

mhkarimi1383 replied:
We should make unmounts to use lock, so it will wait

tiagolobocastro commented on rawfile/bd2fs.py:
```diff
@@ -285,24 +285,49 @@ def ControllerExpandVolume(self, request, context):
+                    for mp in mountpoints:
+                        run(f"fsfreeze --freeze {mp}", check=True)
+                Path(f"{snapshots_path}/{request.name}.img.creating").touch()
+                cmd = f"cp --sparse=always --reflink={'always' if copy_on_write else 'never'}"
```
presumably this will take up a long time and cause some issues with csi retrying?
Should have some kind of "copy manager" which could detect an existing copy is in-progress and return an appropriate error?

tiagolobocastro replied:
Also there's no progress report this way, can anything be done like for example, add annotation every X GiB copied?

mhkarimi1383 replied:
Yeah we should somehow track it

tiagolobocastro commented on rawfile/protos/internal.proto:
```diff
@@ -1,7 +1,10 @@
+
 service Internal {
   rpc ExpandRawFile (ExpandRawFileRequest) returns (ExpandRawFileResponse);
+  rpc GetRawFile (GetRawFileRequest) returns (GetRawFileResponse);
```
Must have missed it, when is this used?

mhkarimi1383 replied:
Clone and migrate between nodes

tiagolobocastro replied:
Is that on this PR already or is this just adding the infra for it?

mhkarimi1383 replied:
Will be implemented here

tiagolobocastro replied:
Could you add some comments here explaining what it is for?

mhkarimi1383 commented:
@mergify rebase

mhkarimi1383 commented:
> @tiagolobocastro With this we can do snapshots in background I think :)
> 
> So if snapshot creation takes longer we won't get timeout errors

@tiagolobocastro What do you think about this

I can add some background workers for handling this using threads

tiagolobocastro commented:
> > @tiagolobocastro With this we can do snapshots in background I think :)
> > So if snapshot creation takes longer we won't get timeout errors
> 
> @tiagolobocastro What do you think about this
> 
> I can add some background workers for handling this using threads

Yeah possibly, with some way of querying if existing task already in progress? Is there a csi error for this scenario or is it expected to just set a huge timeout value?

Btw I suggest you split this into PRs. Ex start with the reflink only as we don't need to worry about long copies with that right?

mhkarimi1383 commented:
> > > @tiagolobocastro With this we can do snapshots in background I think :)
> > > So if snapshot creation takes longer we won't get timeout errors
> > 
> > 
> > @tiagolobocastro What do you think about this
> > I can add some background workers for handling this using threads
> 
> Yeah possibly, with some way of querying if existing task already in progress? Is there a csi error for this scenario or is it expected to just set a huge timeout value?
> 
> Btw I suggest you split this into PRs. Ex start with the reflink only as we don't need to worry about long copies with that right?

I will try to split PRs I will make it to run in background and track progress somehow

mhkarimi1383 commented:
I'm working on a snapshot manager that could handle graceful shutdown

mhkarimi1383 commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,98 @@
+                )
+
+            finally:
+                Path(f"{snap_path}.creating").unlink(missing_ok=True)
```
We need to also make to not mark failed ones as read to use

mhkarimi1383 commented on rawfile/utils/task_manager.py:
```diff
@@ -0,0 +1,119 @@
+    retry_count: int
+
+
+class TaskManager:
```
A global shared mux lock is required and some duplicate codes should be removed

mhkarimi1383 commented:
Something is strange is that external snapshotter is not calling `ListSnapshot` and snapshots are not getting ready

mhkarimi1383 commented:
Also `ControllerGetCapabilities` response is empty


```
2025-09-25 09:12:33.194 | INFO     | utils.logs:wrap:130 - GRPC Server Access Log | {'handler': 'ControllerGetCapabilities', 'request': , 'starttime': datetime.datetime(2025, 9, 25, 9, 12, 33, 194573), 'response': , 'latency': datetime.timedelta(microseconds=48), 'endtime': datetime.datetime(2025, 9, 25, 9, 12, 33, 194621), 'success': True}
```

mhkarimi1383 commented:
@mergify rebase

mhkarimi1383 commented:
We don't have progress percentage tracking, But we have a live update of the size and readiness of the snapshot creation also create time is a real value right now 

And we have a graceful shutdown for that

Also we can move some other parts to use the task manager

mhkarimi1383 commented:
I have to also change the snapshot list and remove to use the new manager

Then we should add a volume manager too I think 🤔 
Since restore of large snapshots may take longer than gRPC timeout

tiagolobocastro commented on rawfile/bd2fs.py:
```diff
@@ -284,22 +289,36 @@ def ControllerExpandVolume(self, request, context):
+                        source_volume_id=snapshot.volume_id,
+                        ready_to_use=snapshot.ready,
+                        creation_time=Timestamp(
+                            seconds=int(snapshot.creation_time),
```
this is implemented differently below

mhkarimi1383 replied:
This is for handling snapshot creation in background, for progress tracking external snapshotter sends multiple snapshot create multiple times, I will check that if the snapshot already exists I will return the existing one's information

tiagolobocastro replied:
I had meant the creation_time only, seems to be set differently though perhaps because we're taking i from different sources?

mhkarimi1383 replied:
Yeah, initially it will be fake, but later on it will get changed to the actual ctime of the file

tiagolobocastro commented on rawfile/bd2fs.py:
```diff
@@ -308,7 +327,7 @@ def CreateSnapshot(self, request: csi_pb2.CreateSnapshotRequest, context):
                         seconds=creation_time_ns // nano, nanos=creation_time_ns % nano
                     ),
-                    ready_to_use=True,
+                    ready_to_use=False,
```
If we're using copy_on_write isn't the snapshot instance, in which can we can return snapshot as ready?

mhkarimi1383 replied:
We are delighting the entire process to task manager and we know if the worker has picked up the job or not

tiagolobocastro replied:
Sorry didn't get you?

mhkarimi1383 replied:
Even using CoW, we are doing it in the background using the background worker, So we don't know if the task manager has picked up the job or not

tiagolobocastro replied:
Yeah that's my doubt, do we need copy manager if we're using cow? 
Otherwise we could have a small wait with timeout in case it's cow?

mhkarimi1383 replied:
Okay, we can wait for 30s for that in the foreground

mhkarimi1383 replied:
Done.

tiagolobocastro commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,115 @@
+            loop_devs = attached_loops(file.as_posix())
+            snap_path = self._get_snapshot_path(volume_id, name)
+            snap_path.parent.mkdir(parents=True, exist_ok=True)
+            Path(f"{snap_path}.creating").touch()
```
Should we add metadata files for snapshots as well rather than just looking at img files?

mhkarimi1383 replied:
By metadata files, we will look at img files later on too, since we need the exact size of the file

mhkarimi1383 replied:
So I think it is not necessary

tiagolobocastro commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,115 @@
+                        run(f"fsfreeze --freeze {mp}", check=True)
+
+                reflink = "always" if copy_on_write else "never"
+                cmd = f"cp --sparse=always --reflink={reflink} {file} {snap_path}"
```
btw seems cp returns not supported with reflink=always if the fs doesn't support it, could we make use of this to determine cow?

mhkarimi1383 replied:
If that's the case we can do a fake cp command to determine that on startup, and save it into memory

tiagolobocastro replied:
yes that would be neat :)

mhkarimi1383 replied:
I will make it to determine that by default, but users are able to override that

mhkarimi1383 replied:
Done.

tiagolobocastro commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,115 @@
+            finally:
+                if freeze_fs and mountpoints:
+                    for mp in mountpoints:
+                        run(f"fsfreeze --unfreeze {mp}", check=True)
```
in case of a crash who unfreezes, is it the task manager by retrying the create again?
and if so, what happens if snapshot is deleted meanwhile?

mhkarimi1383 replied:
We have graceful shutdown, but we could add an static job on startup to unfreeze everything before starting the task manager

tiagolobocastro replied:
And also what happens if we receive another csi call whilst we have a copy manager?
Ex: 
- snapshot may be taken whilst volume is unpublished, then we copy... and then we receive a controllerPublish
- snapshot is taken whilst volume is staged on a node, then we receive a nodePublish or nodeUnpublish
- etc

mhkarimi1383 replied:
We are using volume locks in all of them So other Operations are blocked when doing snapshot

tiagolobocastro replied:
What if the pod restarts?

mhkarimi1383 replied:
a static unfreeze on startup should fix that, that's the only idea that I have

mhkarimi1383 replied:
Why doing FSFreeze when we can Set entire blockdev to be RO?
It is easier to manage and also `attach_loop` will set them to RW automatically

mhkarimi1383 replied:
@tiagolobocastro the only this that is left is this one :)

tiagolobocastro replied:
> Why doing FSFreeze when we can Set entire blockdev to be RO?

Not sure if that works, IIRC there were some pitfalls with setting the device as RO
Also we need to ensure the data if flushed to disk before we can start copying

> a static unfreeze on startup should fix that, that's the only idea that I have

That needs coordination with copy, as if we republish then we have to start copying snapshot back from scratch

mhkarimi1383 replied:
We are already doing it

Task manager will retry

(Only on graceful shutdowns)
I will add a checker for snapshots to retry stuck ones on startup

mhkarimi1383 replied:
I have handled crashes during snapshot creation

I think it is better to handle all of the stuck tasks, So we will have safe clones later on (I will refactor) :)

mhkarimi1383 commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,141 @@
+            try:
+                if freeze_fs and mountpoints:
+                    for mp in mountpoints:
+                        run(f"fsfreeze --freeze {mp}", check=True)
```
This should be added to filesystem interface since it is different in some filesystems (like xfs)

mhkarimi1383 commented:
@mergify rebase

mhkarimi1383 commented:
I want to deprecate and remove filesystem level snapshots :)

tiagolobocastro commented on rawfile/utils/remote.py:
```diff
@@ -48,21 +49,37 @@ def init_rawfile(volume_id, size, thin_provision=False):
         utils.rawfile.patch_metadata(
             volume_id,
             {
                 "schema_version": LATEST_SCHEMA_VERSION,
```
Do we need to bump the version?

mhkarimi1383 replied:
No, because we are falling back to default when getting new keys

tiagolobocastro replied:
Alright

tiagolobocastro commented on rawfile/utils/rawfile.py:
```diff
@@ -65,6 +69,12 @@ def destroy(volume_id, dry_run=True):
         Path(img_file(volume_id)).unlink(missing_ok=True)
         Path(meta_file(volume_id)).unlink(missing_ok=True)
         Path(lock_file(volume_id)).unlink(missing_ok=True)
+        for each in snapshots_dir(volume_id).glob("*"):
```
Deleting a volume should not invalidate the snapshots
In case of cow I guess we'll have to "fake" delete the volume, and in case of full copy it shouldn't matter?

mhkarimi1383 replied:
Or we can simply reject volume deletion when we have snapshots I think some other local PVs are doing the same

tiagolobocastro replied:
At least zfs and mayastor succeed the csi call but keep the volume data around

mhkarimi1383 replied:
I think LVM rejects it,

But we will need GC worker to check for them and remove them

tiagolobocastro replied:
We might need to change LVM then, this is also a change we made on ZFS IIRC.
Could we delete it as part of snapshot delete? Ex: when last reference is deleted we delete the volume

mhkarimi1383 replied:
I will make Delete Snapshot to do GC too and check for this condition

tiagolobocastro commented on rawfile/utils/remote.py:
```diff
@@ -48,21 +49,37 @@ def init_rawfile(volume_id, size, thin_provision=False):
+        if snapshot_id:
+            volume_id, name = snapshot_id.rsplit("/", 1)
+            metadata = utils.rawfile.metadata(volume_id)
+            size = max(size, metadata["size"])
```
Shouldn't the size match exactly the snapshot size?

mhkarimi1383 replied:
Idk, I have made that to the maximum size between snapshot and volume request

tiagolobocastro replied:
For example let's say it's a filesystem volume, if the size is greater than volume we'd have to expand the filesystem? Which could be done though might be easier to ensure it's the same for now and expand later

mhkarimi1383 replied:
Currently it will expand it.
I'll make to to not expand if that's preferred

tiagolobocastro replied:
Oh it's already expanding the filesystem? Awesome :D 
Then that's fine :+1:

mhkarimi1383 replied:
Yeah :)

tiagolobocastro commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,152 @@
+                buf = src.read(chunk_size)
+                if not buf:
+                    break
+                dst.write(buf)
```
Could we also make use of cow here?

mhkarimi1383 replied:
Using COW will make the newly created volume depend on the source snapshot

tiagolobocastro replied:
Yeah, this may be ok as a cheap way of checking snapshot contents for example
Ie if copy_on_write property in the SC is true, then we use cow here, otherwise it's a full copy

mhkarimi1383 replied:
What if user deletes the volume/snapshot that is used as source

tiagolobocastro replied:
We would have to keep the data around and delete it when last reference is deleted; this is how we do it on mayastor
But we could do it as subsequent change to keep it simpler here for now

mhkarimi1383 replied:
OK, For that I have to change the GC behavior and make it to also run after snapshot deletion

mhkarimi1383 commented:
@mergify rebase

mhkarimi1383 commented:
@mergify rebase

mhkarimi1383 commented:
@mergify rebase

tiagolobocastro commented on rawfile/filesystem/base.py:
```diff
@@ -272,3 +288,33 @@ def format_and_mount(
+        """
+        try:
+            run(
+                f"fsfreeze --freeze {self.mountpoint}",
```
Missing idempotency handling? IRRC errors are return if fs is already frozen/unfrozen

mhkarimi1383 replied:
I will test it out and fix it

mhkarimi1383 replied:
I will ignore resource unavailable errors since there is no direct way to check

mhkarimi1383 replied:
Done.

tiagolobocastro commented on rawfile/protos/internal.proto:
```diff
@@ -1,20 +1,46 @@
 service Internal {
+  // Expands raw file to new size.
   rpc ExpandRawFile (ExpandRawFileRequest) returns (ExpandRawFileResponse);
+  // Downlaods raw file data and metadata.
```
```suggestion
  // Downloads raw file data and metadata.
```

tiagolobocastro commented on rawfile/protos/internal.proto:
```diff
@@ -1,20 +1,46 @@
+message GetRawFileResponse {
+  // Metadata associated with the raw file.
+  map<string, google.protobuf.Any> metadata = 1;
+  // Raw file data. Present only if with_data was true in the request.
```
nit: throughout should "raw file" be a single work? "rawfile"

```suggestion
  // Rawfile data. Present only if with_data was true in the request.
```

tiagolobocastro commented on rawfile/protos/internal.proto:
```diff
@@ -1,20 +1,46 @@
+  string volume_id = 1;
+  // Whether to include file data in the response.
+  bool with_data = 2;
+  // Optional snapshot ID to get the raw file at a specific snapshot. will create a temporary snapshot if not provided.
```
how do we keep track and gc this temporary snapshot?

mhkarimi1383 replied:
I have just made an infra for that here, but It will be handled just like other snapshots

tiagolobocastro commented on rawfile/utils/rawfile.py:
```diff
@@ -55,16 +55,27 @@ def img_file(volume_id):
     logger.info("Destroying Volume", volume_id=volume_id, dry_run=dry_run)
     if not dry_run:
+        snapshots = list(snapshots_dir(volume_id).glob("*"))
+        if len(snapshots) > 0:
```
Might be worth adding a log here saying something like "Forgetting about volume xxx because it has yyy dependent snapshots" ?

tiagolobocastro replied:
Ah also, should we check if the snapshots are full copy? If they're full copy then we can delete the volume?

mhkarimi1383 replied:
We can delete the img, meta and lock file
But what if the user didn't specify the COW and changed the filesystem, we don't know if that's COW or not

tiagolobocastro replied:
We could use a snapshot json configuration file?

mhkarimi1383 replied:
Or we can save if we are using COW or not in the snapshot annotation itself

tiagolobocastro replied:
Yes perhaps though might be tricky to get that information here
A config file or even a simple file marker would probably be enough

mhkarimi1383 replied:
In volume meta I'll put that if this volume has reflinks attached to it to or not

mhkarimi1383 replied:
Done.

tiagolobocastro commented on rawfile/utils/task_manager.py:
```diff
@@ -0,0 +1,195 @@
+    def save_task(self, task_id: str, task_info: TaskInfo):
+        with self._lock:
+            logger.debug("Saving task", task_id=task_id, task=task_info)
+            with open(self._tasks_store_path, "r+") as tasks_file:
```
would it be safer to have unique files per task rather than a single for all?

mhkarimi1383 replied:
Handling a specific file per task will be complicated, I know we don't need locking of task file in some places anymore, but it will be complicated I think

tiagolobocastro replied:
Yeah possibly, was just wondering about write atomicity, and what happens if the node crashes half-way through the write
Maybe just the fsync for now is fine

mhkarimi1383 replied:
Yeah
I made it to do fsync and we have locks too

tiagolobocastro commented on rawfile/utils/task_manager.py:
```diff
@@ -0,0 +1,195 @@
+                current[task_id] = task_info
+                tasks_file.truncate(0)
+                tasks_file.seek(0)
+                json.dump(current, tasks_file)
```
fsync as well?

tiagolobocastro commented on docs/design/taskmanager.md:
```diff
@@ -0,0 +1,96 @@
+
+### 2. Task Lifecycle
+
+1. Task submission
```
Should we also add a max number of tasks? IE with a max of 16 tasks, if we get another create_snapshot or create_clone then we'd reject it

mhkarimi1383 replied:
We have a limited thread pool, I will do a quick test to see what will happen, they will get queued I think

tiagolobocastro commented on rawfile/rawfile.py:
```diff
@@ -34,15 +38,36 @@ def csi_driver(driver_config: CSIDriverCmd):
     )
     if driver_config.plugin_type == "node":
         migrate_all_volume_schemas()
+        cow_test_src = Path(f"{consts.DATA_DIR}/cow.test.src")
```
maybe move this logic to the filesystem/init.py ?

mhkarimi1383 replied:
Yeah, it should be moved to a better place

mhkarimi1383 replied:
I will add some startup hooks, I will move some parts there

mhkarimi1383 replied:
Done.

tiagolobocastro commented on docs/design/taskmanager.md:
```diff
@@ -0,0 +1,96 @@
+## Error Handling
+
+- Custom `TaskManagerShuttingDown` exception
+- Automatic crash recovery
```
How are tasks synchronized with csi operations after pod or node restart?

mhkarimi1383 replied:
They are already in the task sore json file with their state

tiagolobocastro replied:
As per the other thread, I think the answer is they are not really synchronized beyond they both take the volume lock? As in, there's no specific log for example: csi_request(): if snap in progress do_this else do_that ?
I wonder if we could massage the returned error to csi by saying snapshot in progress or something?

mhkarimi1383 replied:
Duplicate calls is already handled
we are hashing the name and params of the task and we are using it as task id
So it will not rerun duplicate call

Currently to track if snapshot has been done or not, external snapshotter will call CreateSnapshot multiple times

tiagolobocastro commented on rawfile/bd2fs.py:
```diff
@@ -284,55 +288,104 @@ def ControllerExpandVolume(self, request, context):
+        nano = 10**9
+        current = _get_current_snapshot()
+        if not current:
+            self._task_manager.run_task(
```
there's a race condition here during the time between letting go of the `VolLock` and re-acquiring it here on the `run_task::snapshot_manager::create_snapshot` ?

mhkarimi1383 replied:
Here we only need lock while reading meta file, nothing will happen if we create multiple snapshots of a single volume at the same time

tiagolobocastro commented on rawfile/utils/snapshot_manager.py:
```diff
@@ -0,0 +1,156 @@
+    def _get_snapshot_path(self, volume_id: str, name: str) -> Path:
+        return snapshots_dir(volume_id) / f"{name}.img"
+
+    def create_snapshot(
```
in case of pod/node restart do we always retry the tasks ahead of any csi operations? or is there a chance a csi operation could get through?

mhkarimi1383 replied:
There is a chance for that, we will just mark them as failed and retry worker will pick them

tiagolobocastro replied:
We probably need to ensure the CSI operations don't get in the way or we'd have to make sure we unfreeze the filesystem on mount, because the fs will still be frozen if you unmount/remount for example.

mhkarimi1383 replied:
We can make it unfreeze all on startup

or we can say that this is expected while snapshot creation is in progress

tiagolobocastro replied:
If we don't unfreeze but accept the csi nodeStage call, we risk being frozen forever if not cow, because in that case snapshot task would fail with `SnapshotCreateVolumeInUse`?

mhkarimi1383 replied:
If we have enabled freeze it we not fail with SnapshotCreateVolumeInUse

tiagolobocastro replied:
oh I see, I misread the code below
then the app would just get stuck until the snapshot is taken, I think this is fine then?

mhkarimi1383 replied:
Yep it is expected

tiagolobocastro reviewed (commented):
Have you tested a large copy with cow off and restart the pod mid-way? I suspect it will fail as we're not handling fsfreeze idempotency

tiagolobocastro reviewed (approved):
Only fsfreeze idempotency and the volume deletion thingy outstanding, otherwise this is looking good

mhkarimi1383 commented:
I will also make it to reject block volumes since we can't do fsfreeze on them

mhkarimi1383 commented:
@tiagolobocastro 

FsFreeze problems fixed

tiagolobocastro commented on rawfile/filesystem/base.py:
```diff
@@ -272,3 +289,53 @@ def format_and_mount(
+                capture_output=True,
+            ).stdout.decode()
+        except subprocess.CalledProcessError as e:
+            if "Device or resource busy" in e.stderr.decode().strip():
```
we could also replace this run with an ioctl and then check the codes

mhkarimi1383 replied:
I have read the fsfreeze code itself, and fsfreeze has no return code related to this

mhkarimi1383 replied:
the ioctl will return EBUSY if already frozen, according to https://www.man7.org/linux/man-pages/man3/errno.3.html it means Device or resource busy

tiagolobocastro replied:
Yes, found the mayastor implementation: https://github.com/openebs/mayastor-control-plane/blob/bdb6a11f0733f0134e3db188b9354a7e2ca286db/control-plane/csi-driver/src/bin/node/fsfreeze/bin/ioctl.rs#L31

mhkarimi1383 replied:
So, No need to use ioctl here I think

tiagolobocastro reviewed (approved):
Cool, once this gets in I'll give this a whack
Wonder if there's some other existing csi e2e tests we could leverage for e2e testing

mhkarimi1383 commented:
> Cool, once this gets in I'll give this a whack
> Wonder if there's some other existing csi e2e tests we could leverage for e2e testing

I this we could enable e2e tests for snapshot and restores

mhkarimi1383 commented:
Also we need snapshot classes in chart too

mhkarimi1383 commented:
I have enabled snapshot restore test in the E2E test

mhkarimi1383 commented:
@tiagolobocastro 

It expect when we remove snapshot content provisener should create it or something like that.

Is that a thing that we should cover?

https://github.com/openebs/rawfile-localpv/actions/runs/18398401445/job/52422116518?pr=213#step:15:7665

This should be handled using the external snapshotter and external provisioner since we are managing snapshot content directly

tiagolobocastro commented:
> @tiagolobocastro
> 
> It expect when we remove snapshot content provisener should create it or something like that.
> 
> Is that a thing that we should cover?
> 
> https://github.com/openebs/rawfile-localpv/actions/runs/18398401445/job/52422116518?pr=213#step:15:7665
> 
> This should be handled using the external snapshotter and external provisioner since we are managing snapshot content directly

No, we need not do anything, the side car should handle it. I tend to think of the content like PV for a PVC - it represents the storage itself but we don't have to deal with it directly.

mhkarimi1383 commented:
> > @tiagolobocastro
> > It expect when we remove snapshot content provisener should create it or something like that.
> > Is that a thing that we should cover?
> > https://github.com/openebs/rawfile-localpv/actions/runs/18398401445/job/52422116518?pr=213#step:15:7665
> > This should be handled using the external snapshotter and external provisioner since we are managing snapshot content directly
> 
> No, we need not do anything, the side car should handle it. I tend to think of the content like PV for a PVC - it represents the storage itself but we don't have to deal with it directly.

I have to disable that test I think

mhkarimi1383 commented:
@mergify rebase

mhkarimi1383 commented:
@tiagolobocastro

I'm almost done with this one, I will make CreateVolume to support cloning and make CreateVolume to work in background in a separated PR then.

If everything is OK I will squash 

note that I have deprecated FileSystem level BTRfs snapshots, We should consider telling users that

```
FileSystem Level snapshots should be removed before upgrade to avoid
conflicts
```

tiagolobocastro commented on rawfile/filesystem/btrfs.py:
Don't we need to leave the deletion around to allow deletion of any existing fs snapshots?

mhkarimi1383 replied:
Hmmm...

In `DeleteSnapshot` rpc we should handle that if snapshot does not exist in the block volumes it should be handled via filesystem level handler

mhkarimi1383 commented:
Users should warn that we will only allow deletion of snapshots

and this functionality will be removed

mhkarimi1383 commented:
@mergify squash

mhkarimi1383 committed:
:sparkles: Complete volume snapshot support and better snapshot management  (#213)
:construction: Adding volume clones and snapshot support

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :sparkles: Add generic snapshotting support

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :sparkles: Add restore snapshot

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Using a special snapshot for migration/clone

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :sparkles: Adding a background task manger and snapshot maanger

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :recycle: Make snapshot create async

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :recycle: Use snapshot manager for all operations

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :sparkles: Wait 30s for snapshot readiness of snapshot in forground

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Determine CoW support on startup

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :bug: Snapshot create only lock volume while getting metadata

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :recycle: Use snapshot manager to restore snapshots

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :loud_sound: Taskmanager Logging

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :pencil2: Fix retriable task logic

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :memo: Internal gRPC connection documents

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Volume removal does not remove snapshots

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Requeue stuck snapshots on startup

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Saving task states

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Mark stuck tasks as failed on startup

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Implement filesystem freeze and unfreeze functionality and improve freeze and unfreeze in snapshot manager

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :memo: Regenerate grpc python to Add documentation comments for InternalServicer

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :bug: Fix copy_on_write param handling

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :green_heart: Enable ssh session (Temporary)

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :memo: Add first version of task manager design doc

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :pencil2: Set sparse auto

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :pencil2: Set sparse auto

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :test_tube: Fixing snapshot tests

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* Update rawfile/protos/internal.proto

Co-authored-by: Tiago Castro <tiagolobocastro@gmail.com>
Signed-off-by: Muhammed Hussain Karimi <info@karimi.dev>

* Update rawfile/protos/internal.proto

Co-authored-by: Tiago Castro <tiagolobocastro@gmail.com>
Signed-off-by: Muhammed Hussain Karimi <info@karimi.dev>

* :pencil2: FSync tasks store after saving changes

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :building_construction: Better startup checks

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :pencil2: Handle already frozen and unfreezed filesystems

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :bug: Disallow fsfreeze on block volumes

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Remove volume if depend snapshots are not CoW

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :white_check_mark: Add snapshot restore test in E2E

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :rocket: Add snapshot stuff to helm chart

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :boom: Deprecate filesystem level snapshots

FileSystem Level snapshots should be removed before upgrade to avoid
conflicts

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :wastebasket: Fallback snapshot delete to older implementation, To allow FS Level snapshot deletion

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 commented:
@mergify queue
